File: 2025/01/go/sol-p2.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
)

func main() {
	log.SetFlags(0) // Disable date and time logging
	if _, isDebug := os.LookupEnv("DEBUG"); !isDebug {
		log.SetOutput(io.Discard)
	}

	if len(os.Args) != 2 {
		log.Fatal("Error: Usage: go run sol-p2.go file")
	}

	fileName := os.Args[1]
	fmt.Println("Input file:", fileName)

	file, err := os.Open(fileName)
	if err != nil {
		log.Fatal("Failed to read file:", err)
	}

	curPosition := 50
	countZero := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		instruction := scanner.Text()
		log.Println(instruction)
		
		direction := instruction[:1]
		distance, err := strconv.Atoi(instruction[1:]) 
		if err != nil {
			log.Fatal("Failed to parse distance:", instruction[1:])
		}
		log.Println(direction, distance)

		countZero += distance / 100
		distance %= 100

		if distance == 0 {
			continue
		}

		if direction == "L" {
			distance = -distance
		}
		
		newPosition := (curPosition + distance + 100) % 100
		log.Println(curPosition, distance, newPosition)

		if newPosition == 0 || (distance > 0 && newPosition < curPosition) || (distance < 0 && newPosition > curPosition && curPosition != 0) {
			log.Println("INC")
			countZero++
		}

		curPosition = newPosition
	}

	file.Close()

	fmt.Println("Password:", countZero)
}

File: 2025/01/python/sol-p2.py
import os
import sys

def debug(*kargs):
    if os.getenv("DEBUG") == "1":
        print(*kargs)


def main():
    assert len(sys.argv) == 2, "Usage: python3 sol-p2.py filename"
    
    cur_position = 50
    count_zero = 0

    debug("Input file:", sys.argv[1])

    with open(sys.argv[1], "r") as input_file:
        for instruction in input_file:
            instruction = instruction.strip()
            debug(instruction)

            direction = instruction[0]
            distance = int(instruction[1:])
            debug(direction, distance)

            count_zero += distance // 100
            distance = distance % 100
            
            if distance == 0:
                continue

            if direction == 'L':
                distance = -distance
            
            new_position = (cur_position + distance + 100) % 100
            debug(cur_position, distance, new_position)

            if new_position == 0 or (distance > 0 and new_position < cur_position) or (distance < 0 and new_position > cur_position and cur_position != 0):
                count_zero += 1

            cur_position = new_position

    print("Password: ", count_zero)


if __name__ == "__main__":
    main()

